tests/vmpreset_test.go

```diff
@@ -0,0 +1,218 @@
+	"kubevirt.io/kubevirt/tests"
+)
+
+var _ = Describe("VMPreset", func() {
+
+	dockerTag := os.Getenv("docker_tag")
+	if dockerTag == "" {
+		dockerTag = "latest"
+	}
+
+	flag.Parse()
+
+	virtClient, err := kubecli.GetKubevirtClient()
+	tests.PanicOnError(err)
+
+	var vm *v1.VirtualMachine
+	var memoryPreset *v1.VirtualMachinePreset
+	var cpuPreset *v1.VirtualMachinePreset
+
+	flavorKey := fmt.Sprintf("%s/flavor", v1.GroupName)
+	memoryFlavor := "memory-test"
+	memory, _ := resource.ParseQuantity("128M")
+
+	cpuFlavor := "cpu-test"
+	cores := 7
+
+	BeforeEach(func() {
+		tests.BeforeTestCleanup()
+		vm = tests.NewRandomVMWithEphemeralDisk("kubevirt/alpine-registry-disk-demo:devel")
+		vm.Labels = map[string]string{flavorKey: memoryFlavor}
+
+		selector := k8smetav1.LabelSelector{MatchLabels: map[string]string{flavorKey: memoryFlavor}}
+		memoryPreset = &v1.VirtualMachinePreset{
+			ObjectMeta: k8smetav1.ObjectMeta{GenerateName: "test-memory-"},
+			Spec: v1.VirtualMachinePresetSpec{
+				Selector: selector,
+				Domain: &v1.DomainSpec{
+					Resources: v1.ResourceRequirements{Requests: k8sv1.ResourceList{
+						"memory": memory}},
+				},
+			},
+		}
+
+		selector = k8smetav1.LabelSelector{MatchLabels: map[string]string{flavorKey: cpuFlavor}}
+		cpuPreset = &v1.VirtualMachinePreset{
+			ObjectMeta: k8smetav1.ObjectMeta{GenerateName: "test-cpu-"},
+			Spec: v1.VirtualMachinePresetSpec{
+				Selector: selector,
+				Domain: &v1.DomainSpec{
+					CPU: &v1.CPU{Cores: uint32(cores)},
+				},
+			},
+		}
+	})
+
+	Context("Preset Matching", func() {
+
+		It("Should be accepted on POST", func() {
+			err := virtClient.RestClient().Post().Resource("virtualmachinepresets").Namespace(tests.NamespaceTestDefault).Body(memoryPreset).Do().Error()
+			Expect(err).To(BeNil())
+		})
+
+		It("Should reject a second submission of a VMPreset", func() {
+			// This test requires an explicit name or the resources won't conflict
+			memoryPreset.Name = "test-preset"
+			err := virtClient.RestClient().Post().Resource("virtualmachinepresets").Namespace(tests.NamespaceTestDefault).Body(memoryPreset).Do().Error()
+			Expect(err).ToNot(HaveOccurred())
+
+			waitForPreset(virtClient)
+
+			b, err := virtClient.RestClient().Post().Resource("virtualmachinepresets").Namespace(tests.NamespaceTestDefault).Body(memoryPreset).DoRaw()
+			Expect(err).To(HaveOccurred())
+			status := k8smetav1.Status{}
+			err = json.Unmarshal(b, &status)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("Should return 404 if VMPreset does not exist", func() {
+			b, err := virtClient.RestClient().Get().Resource("virtualmachinepresets").Namespace(tests.NamespaceTestDefault).Name("wrong").DoRaw()
+			Expect(err).To(HaveOccurred())
+			status := k8smetav1.Status{}
+			err = json.Unmarshal(b, &status)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(status.Code).To(Equal(int32(http.StatusNotFound)))
+		})
+
+		It("Should reject presets that conflict with VM settings", func() {
```

Comment:
Not sure if I get it correctly, but we should actually just raise events and annotate the VM that the preset was applied, even if no value was taken over (VM has precedence over preset values)

Reply:
so, annotation essentially just means that the preset was processed in this case